File: src/main/java/olivermakesco/de/refmagic/recipe/AltarRecipe.java
package olivermakesco.de.refmagic.recipe;

import net.minecraft.item.ItemStack;
import net.minecraft.recipe.Ingredient;
import net.minecraft.recipe.Recipe;
import net.minecraft.recipe.RecipeSerializer;
import net.minecraft.recipe.RecipeType;
import net.minecraft.util.Identifier;
import net.minecraft.util.registry.Registry;
import net.minecraft.world.World;
import olivermakesco.de.refmagic.Mod;
import olivermakesco.de.refmagic.block.entity.AltarTableBlockEntity;
import olivermakesco.de.refmagic.recipe.serializer.AltarRecipeSerializer;
import olivermakesco.de.refmagic.registry.RefinedMagicRecipes;

import java.util.Arrays;

public record AltarRecipe(Identifier id, Ingredient catalyst, Ingredient[] inputs,
                          ItemStack result) implements Recipe<AltarTableBlockEntity> {

    public static Identifier ID = Mod.id("altar");

    @Override
    public boolean matches(AltarTableBlockEntity inventory, World world) {
        if (!catalyst.test(inventory.catalyst)) {
            return false;
        }
        boolean[] met = new boolean[4];
        for (int i = 0; i < inventory.size(); i++) {
            ItemStack stack = inventory.getStack(i);
            for (int j = 0; j < inputs.length; j++) {
                if (!met[j]) {
                    if (inputs[j].test(stack)) {
                        met[j] = true;
                    }
                }
            }
        }
        for (boolean b : met) {
            if (!b) {
                return false;
            }
        }
        return true;
    }

    @Override
    public ItemStack craft(AltarTableBlockEntity inventory) {
        return result.copy();
    }

    @Override
    public boolean fits(int width, int height) {
        return width < 4 && height == 1;
    }

    @Override
    public ItemStack getOutput() {
        return result;
    }

    @Override
    public Identifier getId() {
        return id;
    }

    @Override
    public RecipeSerializer<?> getSerializer() {
        return AltarRecipeSerializer.INSTANCE;
    }

    @Override
    public RecipeType<?> getType() {
        return Type.INSTANCE;
    }

    public static class Type implements RecipeType<AltarRecipe> {
        public static final Type INSTANCE = new Type();
    }

    public static void register() {
        Registry.register(Registry.RECIPE_SERIALIZER, ID, AltarRecipeSerializer.INSTANCE);
    }
}

File: src/main/kotlin/olivermakesco/de/refmagic/recipe/AltarRecipe.kt
package olivermakesco.de.refmagic.recipe

import net.minecraft.item.ItemStack
import net.minecraft.recipe.Ingredient
import net.minecraft.recipe.Recipe
import net.minecraft.recipe.RecipeSerializer
import net.minecraft.recipe.RecipeType
import net.minecraft.util.Identifier
import net.minecraft.util.collection.DefaultedList
import net.minecraft.util.registry.Registry
import net.minecraft.world.World
import olivermakesco.de.refmagic.Mod
import olivermakesco.de.refmagic.block.entity.AltarTableBlockEntity
import olivermakesco.de.refmagic.recipe.serializer.AltarRecipeSerializer

data class AltarRecipe(
    @JvmField
    val id: Identifier,
    val catalyst: Ingredient,
    val inputs: Array<Ingredient?>,
    val result: ItemStack
) : Recipe<AltarTableBlockEntity> {
    override fun matches(inventory: AltarTableBlockEntity, world: World?): Boolean {
        if (!catalyst.test(inventory.catalyst)) return false
        val met = BooleanArray(4)
        for (i in 0 until inventory.size()) for (j in inputs.indices) if (!met[j]) met[j] = inputs[j]?.test(
            inventory.getStack(i)
        )!!
        for (b in met) if (!b) return false
        return true
    }

    override fun craft(inventory: AltarTableBlockEntity): ItemStack = result.copy()

    override fun fits(width: Int, height: Int): Boolean = width < 4 && height == 1

    override fun getOutput(): ItemStack = result

    override fun getId(): Identifier = ID

    override fun getIngredients(): DefaultedList<Ingredient> {
        val list = DefaultedList.of<Ingredient>()
        list.addAll(listOf(*inputs))
        return list
    }

    override fun getSerializer(): RecipeSerializer<*> = AltarRecipeSerializer.INSTANCE

    override fun getType(): RecipeType<*> = Type.INSTANCE

    companion object {
        var ID = Mod.id("altar")
        class Type : RecipeType<AltarRecipe?> {
            companion object {
                val INSTANCE: Type = Type()
            }
        }
        fun getTypeInstance(): Type = Type.INSTANCE
        fun register() {
            Registry.register(
                Registry.RECIPE_SERIALIZER,
                ID,
                AltarRecipeSerializer.INSTANCE
            )
        }
    }
}
